### PR TITLE
Provide centralised heatmap state through context

### PR DESCRIPTION
**Pull request by axelboc**
I use a lib called [easy-peasy](https://easy-peasy.now.sh/), more specifically [createContextStore](https://easy-peasy.now.sh/docs/api/create-context-store.html). It's built on top of Redux while providing a simple API for defining actions, computed properties, etc. and for accessing the state with hooks.

I've also refactored `IndexAxis` a little to use VX's `AxisLeft` and `AxisRight` components instead of `Axis`, which simplifies a couple of things.

### CODE REVIEW

File: src/h5web/visualizations/heatmap/IndexAxis.tsx [hunk 1]
```diff
@@ -16,31 +16,33 @@ interface Props {
 
 function IndexAxis(props: Props): JSX.Element {
   const { orientation, numberPixels, className } = props;
+
   const [divRef, { width, height }] = useMeasure();
   const isLeftAxis = orientation === 'left';
+  const Axis = isLeftAxis ? AxisLeft : AxisBottom;
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/IndexAxis.tsx:22]: This is the line that allows rendering one component or the other conditionally.

File: src/h5web/visualizations/heatmap/IndexAxis.tsx [hunk 2]
```diff
@@ -16,31 +16,33 @@ interface Props {
...
-          tickLabelProps={(value, index) =>
-            isLeftAxis
-              ? { className: styles.tick, dx: '-0.25em' }
-              : { className: styles.tick, dy: '0.25em' }
-          }
+          numTicks={numTicks}
+          tickClassName={styles.tick}
+          tickComponent={({ formattedValue, ...tickProps }) => (
+            <text {...tickProps}>{formattedValue}</text>
+          )}
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/IndexAxis.tsx:45]: VX rend son propre composant `Text` par défaut, qui lui même rend un élement `svg`. On se retrouve donc avec `svg > g > svg > text`, ce qui n'est pas génail. `tickComponent` permet d'obtenir `svg > g > text` comme dans l'exemple "Axis" de VX.

File: src/h5web/visualizations/heatmap/interpolators.ts [hunk 3]
```diff
@@ -38,7 +38,7 @@ import {
 
 export type D3Interpolator = (t: number) => string;
 
-export const INTERPOLATORS: Record<string, D3Interpolator> = {
+export const INTERPOLATORS = {
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/interpolators.ts:41]: Avec ce typage, `ColorMap` se retrouvait à valoir `string` au lieu de `'BrBG' | 'PRGn' | ...`. C'est un serpent qui se mord la queue; on ne peut pas avoir un typage fort partout sans définir un enum en plus d'`INTERPOLATORS`.

File: src/h5web/visualizations/heatmap/store.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  }),
+});
+
+export function useHeatmapState(): State<HeatmapState> {
+  return HeatmapStore.useStoreState(state => state);
+}
+
+export function useHeatmapActions(): Actions<HeatmapState> {
+  return HeatmapStore.useStoreActions(state => state);
+}
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/store.ts:23]: These two hooks are just sugar. We could also use `HeatmapStore.useStoreState` and `HeatmapStore.useStoreACtions` directly in our components:

```ts
const colorMap = HeatmapStore.useStoreState(state => state.colorMap);
```

File: src/h5web/visualizations/heatmap/store.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+interface HeatmapState {
+  colorMap: ColorMap;
+  setColorMap: Action<HeatmapState, ColorMap>;
+}
+
+export const HeatmapStore = createContextStore<HeatmapState>({
+  colorMap: 'Magma',
+  setColorMap: action((state, colorMap) => {
+    state.colorMap = colorMap;
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/store.ts:13]: We mutate the state. easy-peasy makes this operation safe, as explained here: https://easy-peasy.now.sh/docs/tutorial/using-actions-to-update-state.html -- unfortunately, ESLint doesn't like this, which is why I disable a rule at the top of the file.

File: src/h5web/visualizations/heatmap/store.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import { createContextStore, Action, State, action, Actions } from 'easy-peasy';
+import { ColorMap } from './interpolators';
+
+interface HeatmapState {
+  colorMap: ColorMap;
+  setColorMap: Action<HeatmapState, ColorMap>;
+}
+
+export const HeatmapStore = createContextStore<HeatmapState>({
+  colorMap: 'Magma',
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/store.ts:11]: Note that by removing the record type from `INTERPOLATORS`, I now get proper auto-completion for `colorMap` here.

File: src/h5web/visualizations/heatmap/store.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    state.colorMap = colorMap;
+  }),
+});
+
+export function useHeatmapState(): State<HeatmapState> {
+  return HeatmapStore.useStoreState(state => state);
+}
+
+export function useHeatmapActions(): Actions<HeatmapState> {
+  return HeatmapStore.useStoreActions(state => state);
```
Comment (loichuder) [inline src/h5web/visualizations/heatmap/store.ts:22]: Should not it be `HeatmapStore.useStoreActions(actions => actions);` for consistency ?

Comment (axelboc) [reply src/h5web/visualizations/heatmap/store.ts:22]: yes, t'as raison, je fix dans ma nouvelle branche pour me simplifier la vie

**Pull request merged**